store/src/java/com/zimbra/cs/mime/MimeUtil.java

```diff
@@ -0,0 +1,179 @@
+            b64 = false;
+        } else {
+            throw new UnsupportedEncodingException(
+                    "Unknown transfer encoding: " + encoding);
+        }
+
+        StringBuffer outb = new StringBuffer(); // the output buffer
+        doEncode(string, b64, jcharset,
+                // As per RFC 2047, size of an encoded string should not
+                // exceed 75 bytes.
+                // 7 = size of "=?", '?', 'B'/'Q', '?', "?="
+                75 - 7 - charset.length(), // the available space
+                "=?" + charset + "?" + encoding + "?", // prefix
+                true, encodingWord, outb);
+
+        return outb.toString();
+    }
+
+    /*
+     * The following two properties allow disabling the fold()
+     * and unfold() methods and reverting to the previous behavior.
+     * They should never need to be changed and are here only because
+     * of my paranoid concern with compatibility.
+     */
+    private static final boolean foldEncodedWords =
+            PropUtil.getBooleanSystemProperty("mail.mime.foldencodedwords", false);
+
+    /** (non-Javadoc)
+     * @see javax.mail.internet.MimeUtility.doEncode(String string, boolean b64, String jcharset, int avail, String prefix, boolean first, boolean encodingWord, StringBuffer buf)
+     */
+    private static void doEncode(String string, boolean b64,
+            String jcharset, int avail, String prefix,
+            boolean first, boolean encodingWord, StringBuffer buf)
+                    throws UnsupportedEncodingException {
+
+        // First find out what the length of the encoded version of
+        // 'string' would be.
+        byte[] bytes = string.getBytes(jcharset);
+        int len;
+        if (b64) {
+            // "B" encoding
+            len = BEncoderStream.encodedLength(bytes);
+        } else {
+            // "Q"
+            len = QEncoderStream.encodedLength(bytes, encodingWord);
+        }
+
+        int size;
+        if ((len > avail) && ((size = string.length()) > 1)) {
+            // If the length is greater than 'avail', split 'string'
+            // into two and recurse.
+            doEncode(string.substring(0, size/2), b64, jcharset,
+                    avail, prefix, first, encodingWord, buf);
+            doEncode(string.substring(size/2, size), b64, jcharset,
+                    avail, prefix, false, encodingWord, buf);
+        } else {
+            // length <= than 'avail'. Encode the given string
+            ByteArrayOutputStream os = new ByteArrayOutputStream();
+            OutputStream eos; // the encoder
+            if (b64) {
+                // "B" encoding
+                eos = new BEncoderStream(os);
+            } else {
+                // "Q" encoding
+                eos = new QEncoderStream(os, encodingWord);
+            }
+
+            try { // do the encoding
+                eos.write(bytes);
+                eos.close();
+            } catch (IOException ioex) { }
```

Comment:
log or throw exception here?

Reply:
The "eos" (as wellas "os") virtually does not throw the IOException; the write() method could throw the NullPointerException or IndexOutOfBoundsException, but the root cause of both exceptions has been taken care before.  So, in this code, that is just a formality to catch such exception.